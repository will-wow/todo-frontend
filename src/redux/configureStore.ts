import { applyMiddleware, createStore, Store as ReduxStore } from "redux";
import thunkMiddleware from "redux-thunk";

import rootReducer, { Store } from "./reducers";

export default function configureStore(): ReduxStore<Store> {
  const preloadedState = {};

  const store = createStore(
    rootReducer,
    preloadedState,
    applyMiddleware(thunkMiddleware)
  );

  return store;
}
