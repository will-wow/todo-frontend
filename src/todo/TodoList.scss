.TodoList {
  max-width: 400px;
  margin: 0 auto;
  display: flex;
  flex-direction: column;
}
