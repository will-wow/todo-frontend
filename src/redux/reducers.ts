import todo, { TodoStore } from "../todo/redux/reducer";

import { combineReducers } from "redux";

export interface Store {
  todo: TodoStore;
}

const reducers = combineReducers({ todo });

export default reducers;
