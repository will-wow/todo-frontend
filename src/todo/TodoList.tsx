import React from "react";
import { filter } from "lodash";
import { Box, Heading, Flex } from "rebass";

import * as Todo from "./todo";
import TodoItem from "./TodoItem";

import InboxZero from "../user/InboxZero";
import UsernameInput from "../user/UsernameInput";
import Hr from "../theme/Hr";

export interface TodoListProps {
  todoList: Todo.List;
  onTodoChange: (todo: Todo.T) => void;
  onDelete: (todo: Todo.T) => void;
}

const TodoList: React.FC<TodoListProps> = ({
  todoList,
  onTodoChange,
  onDelete
}) => {
  const notDone = filter(todoList, { done: false });
  const done = filter(todoList, { done: true });
  const isAllDone = notDone.length === 1;

  return (
    <Box className="TodoList" mx="auto" style={{ maxWidth: "25rem" }}>
      <Flex mb={2} justifyContent="space-between" alignItems="flex-end">
        <Heading>Todo List for:</Heading>
        <UsernameInput />
      </Flex>

      <Hr borderColor="gray" />

      {isAllDone && <InboxZero />}

      {notDone.map(todo => (
        <TodoItem
          key={todo.uuid || todo.id}
          todo={todo}
          onChange={onTodoChange}
          onDelete={onDelete}
        />
      ))}

      <Heading fontSize={3} mt={2}>
        Done Items
      </Heading>
      {done.map(todo => (
        <TodoItem
          todo={todo}
          key={todo.uuid || todo.id}
          onChange={onTodoChange}
          onDelete={onDelete}
        />
      ))}
    </Box>
  );
};

TodoList.defaultProps = {
  todoList: []
};

export default TodoList;
