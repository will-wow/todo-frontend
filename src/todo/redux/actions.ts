import { isError } from "result-async";

import * as Todo from "../todo";
import * as Api from "../../api";

interface LoadTodoListSuccess {
  type: "loadTodoListSuccess";
  payload: {
    todoList: Todo.List;
  };
}

interface UpdateTodo {
  type: "updateTodo";
  payload: {
    todo: Todo.T;
  };
}

interface CreateTodoSuccess {
  type: "createTodoSuccess";
  payload: {
    todo: Todo.T;
  };
}

interface RemoveTodo {
  type: "removeTodo";
  payload: {
    todo: Todo.T;
  };
}

interface NewTodo {
  type: "newTodo";
}

export type TodoAction =
  | LoadTodoListSuccess
  | UpdateTodo
  | CreateTodoSuccess
  | RemoveTodo
  | NewTodo;

export const loadTodoList = () => async (
  dispatch: React.Dispatch<TodoAction>
) => {
  const response = await Api.get<Todo.List, string>("/");

  if (isError(response)) return null;

  const todoList = response.ok;

  dispatch({ type: "loadTodoListSuccess", payload: { todoList } });
  dispatch({ type: "newTodo" });
};

export const updateTodo = (todo: Todo.T) => async (
  dispatch: React.Dispatch<TodoAction>
) => {
  dispatch({ type: "updateTodo", payload: { todo: todo } });

  if (!Todo.isNew(todo)) {
    return Api.put<Todo.T, string>(`/${todo.id}`, Todo.toApi(todo));
  } else {
    const newTodo = await Api.post<Todo.T, string>("/", Todo.toApi(todo));

    if (isError(newTodo)) return null;

    const updatedTodo = { ...todo, ...newTodo.ok };

    dispatch({ type: "createTodoSuccess", payload: { todo: updatedTodo } });
    dispatch({ type: "newTodo" });
  }
};

export const deleteTodo = (todo: Todo.T) => async (
  dispatch: React.Dispatch<TodoAction>
) => {
  await Api.remove<Todo.T, string>(`/${todo.id}`);

  dispatch({ type: "removeTodo", payload: { todo } });
};
