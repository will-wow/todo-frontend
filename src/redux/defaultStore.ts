import { Store } from "./reducers";

const defaultStore: Store = {
  todo: {
    todoMap: {}
  }
};

export default defaultStore;
