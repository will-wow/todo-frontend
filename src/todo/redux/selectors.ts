import * as Todo from "../todo";
import { Store } from "../../redux/reducers";

export const getTodoList = (state: Store): Todo.List =>
  Todo.mapToList(state.todo.todoMap);
