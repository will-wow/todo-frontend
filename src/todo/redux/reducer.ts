import * as Todo from "../todo";
import { TodoAction } from "./actions";
import { Reducer } from "redux";
import defaultStore from "../../redux/defaultStore";

export interface TodoStore {
  todoMap: Todo.Map;
}

const reducer: Reducer<TodoStore, TodoAction> = (
  state = defaultStore.todo,
  action
) => {
  switch (action.type) {
    case "newTodo": {
      const todo = Todo.build();
      return {
        ...state,
        todoMap: {
          ...state.todoMap,
          [Todo.key(todo)]: todo
        }
      };
    }
    case "loadTodoListSuccess": {
      return {
        ...state,
        todoMap: Todo.listToMap(action.payload.todoList)
      };
    }
    case "updateTodo": {
      const { todo } = action.payload;
      const { todoMap } = state;

      const updatedMap = { ...todoMap, [Todo.key(todo)]: todo };

      return { ...state, todoMap: updatedMap };
    }
    case "createTodoSuccess": {
      const { todo } = action.payload;
      const { todoMap } = state;

      const updatedMap = {
        ...todoMap,
        [Todo.key(todo)]: todo
      };

      return { ...state, todoMap: updatedMap };
    }
    case "removeTodo": {
      const { todo } = action.payload;
      const { todoMap } = state;

      const updatedMap = { ...todoMap };
      delete updatedMap[Todo.key(todo)];

      return { ...state, todoMap: updatedMap };
    }
    default: {
      return state;
    }
  }
};

export default reducer;
