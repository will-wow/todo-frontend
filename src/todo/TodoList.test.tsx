import React from "react";
import { shallow, ShallowWrapper, mount, ReactWrapper } from "enzyme";

import renderer from "react-test-renderer";
import { createRenderer } from "react-test-renderer/shallow";
import ReactTestUtils from "react-dom/test-utils";

import { render, fireEvent } from "@testing-library/react";

import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";

import TodoList, { TodoListProps } from "./TodoList";

import { Store } from "../redux/reducers";
import defaultStore from "../redux/defaultStore";

const middleware = [thunk];
const mockStore = configureMockStore<Store>(middleware);

const TODO_LIST = [
  {
    id: 1,
    title: "Pending Item",
    done: false,
    createdAt: "2019-01-01"
  },
  {
    id: "",
    uuid: "abc",
    title: "New Item",
    done: false,
    createdAt: "2019-02-01"
  },
  {
    id: 2,
    title: "Done Item",
    done: true,
    createdAt: "2019-01-01"
  }
];

describe("TodoList", () => {
  let props: TodoListProps;

  beforeEach(() => {
    props = {
      todoList: TODO_LIST,
      onTodoChange: jest.fn(),
      onDelete: jest.fn()
    };
  });

  describe("full render", () => {
    describe("react-test-renderer", () => {
      let component: renderer.ReactTestRenderer;

      beforeEach(() => {
        const store = mockStore(defaultStore);

        component = renderer.create(
          <Provider store={store}>
            <TodoList {...props} />
          </Provider>
        );
      });

      it("snapshots", () => {
        expect(component.toJSON()).toMatchSnapshot();
      });
    });
  });

  describe("shallow", () => {
    describe("react-test-renderer", () => {
      let component: React.ReactElement;

      beforeEach(() => {
        const renderer = createRenderer();
        renderer.render(<TodoList {...props} />);
        component = renderer.getRenderOutput();
      });

      it("snapshots", () => {
        expect(component).toMatchSnapshot();
      });
    });

    describe("enzyme", () => {
      let wrapper: ShallowWrapper;

      beforeEach(() => {
        wrapper = shallow(<TodoList {...props} />);
      });

      it("snapshots", () => {
        expect(wrapper.debug()).toMatchSnapshot();
      });

      it("renders all todo items", () => {
        const items = wrapper.find("TodoItem");

        // Correct number of items exist
        expect(items).toHaveLength(TODO_LIST.length);

        // Passing todo list
        expect(items.first().prop("todo")).toEqual(TODO_LIST[0]);
      });

      it("changes todo items", () => {
        wrapper
          .find("TodoItem")
          .first()
          .simulate("change", TODO_LIST[0]);

        expect(props.onTodoChange).toHaveBeenCalledWith(TODO_LIST[0]);
      });
    });
  });
});
