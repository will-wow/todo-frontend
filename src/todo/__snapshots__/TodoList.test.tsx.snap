// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList full render react-test-renderer snapshots 1`] = `
<div
  className="TodoList sc-bdVaJa hTTfNc"
  style={
    Object {
      "maxWidth": "25rem",
    }
  }
>
  <div
    className="sc-bdVaJa sc-bwzfXH dpqGKp"
  >
    <h2
      className="sc-bdVaJa sc-htpNat sc-bxivhb kXNIHy"
      fontSize={4}
      fontWeight="bold"
    >
      Todo List for:
    </h2>
    <input
      className="sc-bdVaJa sc-htpNat sc-htoDjs jplWqw"
      fontSize={1}
      onChange={[Function]}
      placeholder="Enter your name here"
      value="User"
    />
  </div>
  <hr
    className="sc-dnqmqq WskOd"
  />
  <div
    className="sc-bdVaJa TodoItem sc-bwzfXH dKmwgv"
  >
    <input
      className="sc-bdVaJa sc-htpNat TodoItem__input sc-htoDjs jplWqw"
      fontSize={1}
      onChange={[Function]}
      placeholder="Enter a todo item"
      style={
        Object {
          "flexGrow": 1,
          "textDecoration": "",
        }
      }
      value="Pending Item"
    />
    <button
      className="sc-bdVaJa sc-EHOje kturus"
      color="white"
      fontSize="inherit"
      fontWeight="bold"
      onClick={[Function]}
    >
      Done
    </button>
    <button
      className="sc-bdVaJa sc-EHOje gPdMQc"
      color="white"
      fontSize="inherit"
      fontWeight="bold"
      onClick={[Function]}
    >
      X
    </button>
  </div>
  <div
    className="sc-bdVaJa TodoItem sc-bwzfXH dKmwgv"
  >
    <input
      className="sc-bdVaJa sc-htpNat TodoItem__input sc-htoDjs jplWqw"
      fontSize={1}
      onChange={[Function]}
      placeholder="Enter a todo item"
      style={
        Object {
          "flexGrow": 1,
          "textDecoration": "",
        }
      }
      value="New Item"
    />
    <button
      className="sc-bdVaJa sc-EHOje kturus"
      color="white"
      fontSize="inherit"
      fontWeight="bold"
      onClick={[Function]}
    >
      Done
    </button>
    <button
      className="sc-bdVaJa sc-EHOje gPdMQc"
      color="white"
      fontSize="inherit"
      fontWeight="bold"
      onClick={[Function]}
    >
      X
    </button>
  </div>
  <h2
    className="sc-bdVaJa sc-htpNat sc-bxivhb rLXOD"
    fontSize={3}
    fontWeight="bold"
  >
    Done Items
  </h2>
  <div
    className="sc-bdVaJa TodoItem sc-bwzfXH dKmwgv"
  >
    <input
      className="sc-bdVaJa sc-htpNat TodoItem__input sc-htoDjs jplWqw"
      fontSize={1}
      onChange={[Function]}
      placeholder="Enter a todo item"
      style={
        Object {
          "flexGrow": 1,
          "textDecoration": "line-through",
        }
      }
      value="Done Item"
    />
    <button
      className="sc-bdVaJa sc-EHOje isXsaM"
      color="white"
      fontSize="inherit"
      fontWeight="bold"
      onClick={[Function]}
    >
      Undo
    </button>
    <button
      className="sc-bdVaJa sc-EHOje gPdMQc"
      color="white"
      fontSize="inherit"
      fontWeight="bold"
      onClick={[Function]}
    >
      X
    </button>
  </div>
</div>
`;

exports[`TodoList shallow enzyme snapshots 1`] = `
"<styled.div className=\\"TodoList\\" mx=\\"auto\\" style={{...}}>
  <Styled(styled.div) mb={2} justifyContent=\\"space-between\\" alignItems=\\"flex-end\\">
    <Styled(Styled(styled.div)) as=\\"h2\\" m={0} fontSize={4} fontWeight=\\"bold\\">
      Todo List for:
    </Styled(Styled(styled.div))>
    <Connect(UsernameInput) />
  </Styled(styled.div)>
  <styled.hr borderColor=\\"gray\\" borderWidth={1} />
  <TodoItem todo={{...}} onChange={[Function: mockConstructor]} onDelete={[Function: mockConstructor]} />
  <TodoItem todo={{...}} onChange={[Function: mockConstructor]} onDelete={[Function: mockConstructor]} />
  <Styled(Styled(styled.div)) fontSize={3} mt={2} as=\\"h2\\" m={0} fontWeight=\\"bold\\">
    Done Items
  </Styled(Styled(styled.div))>
  <TodoItem todo={{...}} onChange={[Function: mockConstructor]} onDelete={[Function: mockConstructor]} />
</styled.div>"
`;

exports[`TodoList shallow react-test-renderer snapshots 1`] = `
<ForwardRef(styled.div)
  className="TodoList"
  mx="auto"
  style={
    Object {
      "maxWidth": "25rem",
    }
  }
>
  <ForwardRef(Styled(styled.div))
    alignItems="flex-end"
    justifyContent="space-between"
    mb={2}
  >
    <ForwardRef(Styled(Styled(styled.div)))
      as="h2"
      fontSize={4}
      fontWeight="bold"
      m={0}
    >
      Todo List for:
    </ForwardRef(Styled(Styled(styled.div)))>
    <Memo(ConnectFunction) />
  </ForwardRef(Styled(styled.div))>
  <ForwardRef(styled.hr)
    borderColor="gray"
    borderWidth={1}
  />
  <TodoItem
    onChange={[MockFunction]}
    onDelete={[MockFunction]}
    todo={
      Object {
        "createdAt": "2019-01-01",
        "done": false,
        "id": 1,
        "title": "Pending Item",
      }
    }
  />
  <TodoItem
    onChange={[MockFunction]}
    onDelete={[MockFunction]}
    todo={
      Object {
        "createdAt": "2019-02-01",
        "done": false,
        "id": "",
        "title": "New Item",
        "uuid": "abc",
      }
    }
  />
  <ForwardRef(Styled(Styled(styled.div)))
    as="h2"
    fontSize={3}
    fontWeight="bold"
    m={0}
    mt={2}
  >
    Done Items
  </ForwardRef(Styled(Styled(styled.div)))>
  <TodoItem
    onChange={[MockFunction]}
    onDelete={[MockFunction]}
    todo={
      Object {
        "createdAt": "2019-01-01",
        "done": true,
        "id": 2,
        "title": "Done Item",
      }
    }
  />
</ForwardRef(styled.div)>
`;
